### PR TITLE
fix(CrossFilters): Do not reload unrelated filters in global scope

### PR DESCRIPTION
**Pull request by geido**
### SUMMARY
This PR partially accounts for the issue reported on #30061. However, more work is required so that when a scope is applied the frontend can check columns / calculated columns across different datasets to see whether the query should be sent over the backend at all. Working on a follow-up PRs that should solve the issue for both native and cross filters.

### BEFORE

https://github.com/user-attachments/assets/7b91a4a8-2c14-4bfb-84f4-94d82d02ec84

### AFTER

https://github.com/user-attachments/assets/6a85fc52-2c13-4cf4-977f-f9623c0a55f3

### TESTING INSTRUCTIONS
1. In global scope a cross filter should only affect charts of the same dataset

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by betodealmeida — APPROVED** [review]
👍🏼

File: superset-frontend/src/dashboard/util/crossFilters.ts [hunk 1]
```diff
@@ -52,6 +52,20 @@ export const getCrossFiltersConfiguration = (
...
+    charts,
+  ).reduce((acc: Record<string, number[]>, chart) => {
+    if (!chart.form_data) {
+      return acc;
+    }
+    const { datasource } = chart.form_data;
+    if (!acc[datasource]) {
+      acc[datasource] = [];
+    }
+    acc[datasource].push(chart.id);
```
Comment (justinpark) [inline superset-frontend/src/dashboard/util/crossFilters.ts:65]: Using a set would likely make the algorithm more efficient.
```suggestion
    if (!acc[datasource]) {
      acc[datasource] = new Set();
    }
    acc[datasource].add(chart.id);
```

File: superset-frontend/src/dashboard/util/crossFilters.ts [hunk 2]
```diff
@@ -97,10 +111,13 @@ export const getCrossFiltersConfiguration = (
...
         };
       }
+      const chartDataSource = charts[chartId].form_data.datasource;
       chartConfiguration[chartId].crossFilters.chartsInScope =
         isCrossFilterScopeGlobal(chartConfiguration[chartId].crossFilters.scope)
           ? globalChartConfiguration.chartsInScope.filter(
-              id => id !== Number(chartId),
+              id =>
+                id !== Number(chartId) &&
+                chartsByDataSource[chartDataSource].includes(id),
```
Comment (justinpark) [inline superset-frontend/src/dashboard/util/crossFilters.ts:120]: Including optional chaining would likely help prevent undefined errors.

```suggestion
                chartsByDataSource[chartDataSource]?.has(id),
```

**Comment by michael-s-molina** [comment]
> This PR partially accounts for the issue reported on https://github.com/apache/superset/issues/30061. However, more work is required so that when a scope is applied the frontend can check columns / calculated columns across different datasets to see whether the query should be sent over the backed at all. Working on a follow-up PRs that should solve the issue for both native and cross filters.

@geido Do you have a link for the follow-up that will fix the cross filter issue?

**Comment by geido** [comment]
> > This PR partially accounts for the issue reported on https://github.com/apache/superset/issues/30061. However, more work is required so that when a scope is applied the frontend can check columns / calculated columns across different datasets to see whether the query should be sent over the backed at all. Working on a follow-up PRs that should solve the issue for both native and cross filters.
> 
> @geido Do you have a link for the follow-up that will fix the cross filter issue?

Still working on it

**Comment by fmannhardt** [comment]
@geido I am testing the 4.1 branch and realized that our cross filters are not working anymore. Do I read this PR right in that it would entirely disable cross filtering if the data source is not the same, even if shared columns exist? And that this is fixed, also for columns not actually used in the chart, with: https://github.com/apache/superset/pull/30438

**Comment by geido** [comment]
> @geido I am testing the 4.1 branch and realized that our cross filters are not working anymore. Do I read this PR right in that it would entirely disable cross filtering if the data source is not the same, even if shared columns exist? And that this is fixed, also for columns not actually used in the chart, with: #30438

That is a good point @fmannhardt . I updated https://github.com/apache/superset/pull/30438 to make sure we are targeting all possible common columns in all datasources. Let me know if that makes sense to you now.

**Comment by fmannhardt** [comment]
When testing including https://github.com/apache/superset/pull/30438 I still do not see the chart being refreshed. The query is correct and also when looking at the result table it is correct but the chart visualization itself does not show up refreshed.

**Comment by geido** [comment]
> When testing including #30438 I still do not see the chart being refreshed. The query is correct and also when looking at the result table it is correct but the chart visualization itself does not show up refreshed.

@fmannhardt would you mind providing a test case to reproduce this issue on my side, please? That would help.

**Comment by fmannhardt** [comment]
Ok. I will try to provide it. Maybe it had to do with using JINJA templating in a filtering statement.

**Comment by fmannhardt** [comment]
I have tried to reproduce with the latest changes on the other PR and the issue is gone. Thanks.

**Pull request merged**